src/iterators/intersection_iterator.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+#include "intersection_iterator.h"
+#include "empty_iterator.h"
+#include "union_iterator.h"
+#include "index_result.h"
+
+/**************************** Read + SkipTo Helpers ****************************/
+
+static inline bool II_currentIsRelevant(IntersectionIterator *it) {
+  return IndexResult_IsWithinRange(it->base.current, it->max_slop, it->in_order);
+}
+
+static IteratorStatus II_AgreeOnDocId(IntersectionIterator *it) {
+  const t_docId docId = it->base.lastDocId;
+  AggregateResult_Reset(it->base.current);
+  for (uint32_t i = 0; i < it->num_its; i++) {
+    RS_ASSERT(it->its[i]->lastDocId <= docId);
+    if (it->its[i]->lastDocId < docId) {
+      // Advance the iterator to the requested docId
+      IteratorStatus rc = it->its[i]->SkipTo(it->its[i], docId);
+      if (rc != ITERATOR_OK) {
+        if (rc == ITERATOR_EOF) {
+          // Some child iterator reached EOF, so the intersection is also at EOF
+          it->base.atEOF = true;
+        } else if (rc == ITERATOR_NOTFOUND) {
+          it->base.lastDocId = it->its[i]->lastDocId;
```

Comment:
I miss this logic, how are you sure you need to move here? I may miss some logic surrounding this? Maybe a better function naming can provide more context